accepted/2021/winforms/streamline-application-bootstrap.md

```diff
@@ -0,0 +1,362 @@
+
+```
+Name=fontName, Size=size[, Units=units[, GDiCharSet=gdiCharSet[, GdiVerticalFont=boolean]]]
+```
+
+#### Locale aware font configuration
+
+In addition to this, it is possible that [`Font.FamilyName`](https://docs.microsoft.com/dotnet/api/system.drawing.font.fontfamily) may be locale sensitive, i.e. a different font family (and size) could be used in different regions of the world. To address this use case the proposal is to add another MSBuild property `ApplicationFontResx`, that will contain the name of a localizable resource that contains [`Font.FamilyName`](https://docs.microsoft.com/dotnet/api/system.drawing.font.fontfamily) and [`Font.Size`](https://docs.microsoft.com/dotnet/api/system.drawing.font.size) properties for each locale that requires it.
+
+The order of precedence is as follows:
+- `ApplicationFontResx`, if present
+- `ApplicationFont`.
+
+#### Error scenarios
+
+* If the value of `ApplicationFont` resource can't  be parsed, this will result in a compilation error.
+* If `ApplicationFontResx` contains an invalid resource name, this will result in a compilation error.
+* If the value of `ApplicationFontResx` resource can't  be parsed, this will result in a runtime error.
+
+
+
+### How to resolve dpi settings?
+
+At this point an astute Windows Forms developer will say that there are currently 3 different ways to configure dpi settings (which may or may not be mutually exclusive or complimentary):
+* [app.manifest](https://docs.microsoft.com/windows/win32/sbscs/application-manifests) - Common Control related configuration, dealing with it was always a pain. Very likely that a significant part of Windows Forms developers have never heard of it.
+* [app.config ApplicationConfigurationSection](https://docs.microsoft.com/dotnet/framework/configure-apps/file-schema/winforms/windows-forms-add-configuration-element) and [app.config AppContextSwitchOverrides](https://docs.microsoft.com/dotnet/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element).
+* [`Application.SetHighDpiMode(HighDpiMode)`](https://docs.microsoft.com/dotnet/api/system.windows.forms.application.sethighdpimode).
+
+This proposal introduces a 4th way of configuring dpi, and to make it successful it has to (paraphrasing JRR Tolkien):
+
+
+> _One API to rule them all, One API to find them,_<br/>
+> _One API to bring them all and in the ApplicationConfiguration bind them..._
+
+
+The benefit the new approach provides is by facilitating the sharing of the information between the runtime and the designer during the development phase, as well as unifying how dpi settings are configured. This benefit is believed to outweigh the need to remove several lines of code from `Main()` method and adding several MSBuild properties.
+
+### Go further?
+
+At this point an astute reader may ask why no go further and offer an API similar to a ASP.<!-- -->NET Core bootstrap? Wouldn't it be nice to write something like:
+```cs
+[STAThread]
+static void Main()
```

Comment:
Why does the "Go further" still use the old syntax? It seems like there is no value to using "Main" here.

Reply:
Today there is no way to apply the `[STAThread]` to the Main method of a C# program using top level statements. We have a thread with the language design team to discuss options there.

Reply:
You could do something like this 😈

```
await Application.SwitchToSTAThread();
```

Reply:
What would the implementation of that look like?  AFAIK, that requires there to already be an STA thread in the application, which there isn't in this case.  Instead you'd have to create a thread, set it's apartment state, and run the rest of the method as a lambda/local function, which isn't really any better than a Main method.

Reply:
The single statement `await Application.SwitchToSTAThread();` could start a new STA thread and run the rest of the method on it following the await.

Reply:
Evidence:
![image](https://user-images.githubusercontent.com/8040367/121101984-e721b800-c7ca-11eb-9f0a-697330a3c772.png)

<details>
<summary>Click to expand</summary>

```cs
using System;
using System.Runtime.CompilerServices;
using System.Threading;

Console.WriteLine(Thread.CurrentThread.GetApartmentState());
await Application.SwitchToStaThread();
Console.WriteLine(Thread.CurrentThread.GetApartmentState());

public static class Application
{
    public static StaThreadAwaitable SwitchToStaThread() => default;

    public readonly struct StaThreadAwaitable
    {
        public StaThreadAwaiter GetAwaiter() => default;
    }

    public readonly struct StaThreadAwaiter : ICriticalNotifyCompletion
    {
        public bool IsCompleted => Thread.CurrentThread.GetApartmentState() == ApartmentState.STA;

        public void OnCompleted(Action continuation) => UnsafeOnCompleted(continuation);

        public void UnsafeOnCompleted(Action continuation)
        {
            // Real impl would also use existing thread if a message pump has already started

            var thread = new Thread(parameter => ((Action)parameter).Invoke());
            thread.SetApartmentState(ApartmentState.STA);
            thread.Start(parameter: continuation);
        }

        public void GetResult() { }
    }
}
```

</details>

Reply:
> The single statement await `Application.SwitchToSTAThread();` could start a new STA thread

Creating new thread is unnecessary overhead. You can switch the main thread apartment state by doing this as first thing in Main:

```
// It is unfortunate that you have to set it to Unknown first. If there was a good reason, 
// we may be able to change the runtime to make it unnecessary.
Thread.CurrentThread.SetApartmentState(ApartmentState.Unknown);

Thread.CurrentThread.SetApartmentState(ApartmentState.STA);
```

> Today there is no way to apply the [STAThread] to the Main method of a C# program using top level statements.

The attributes on the Main method are legacy. I would not try to complicate the language to support them.

Reply:
We did not respect the attributes on Main in .NET Core 1.0 at all and instead tell people to set the apartment explicitly via a method call. We have added back the support to respect the attributes on Main to make existing (mostly WinForms/WPF) apps work.

Reply:
Great discussion! Just worth noting, the top level application is orthogonal to this design proposal, which is about bridging the gap between the runtime and the designer. 
It doesn't look there's anything that prevents developers from writing Windows Forms apps as top level apps now (modulo some additional boiler plate). This topic is being actively discussed, and it is something we will explore separately.

Reply:
![image](https://user-images.githubusercontent.com/4403806/121298454-3d136e80-c937-11eb-9b60-5ea95a3527ae.png)

Reply:
@jkotas FWIW the app seems to run fine without setting thread apartment states ¯\\_(ツ)\_/¯, though I guess the lack of it may lead to some random failures somewhere else...

Reply:
Windows GDI does not depend on COM at all. It explains why classic Win32 controls work just fine without COM initialized. (In fact, the classic old Win32 apps tried to avoid initializating COM back in the day since COM initialization was a heavy operation that took considerable amount of time and memory.)

The STA appartment has to be only set to use COM and only in some cases. I would expect that you may see bugs with web browser control that is heavily COM-based.

Reply:
I think the file dialogs may also rely on it. There are also single-threaded unit tests involving DevExpress controls that fail if that thread is not an STA thread, and I just assumed this was a general Windows Forms thing. Clipboard operations fail if not on an STA thread.

Reply:
> The STA appartment has to be only set to use COM and only in some cases. I would expect that you may see bugs with web browser control that is heavily COM-based.

The first one I usually hit is clipboard :-)

Reply:
Opened https://github.com/dotnet/winforms/issues/5071 for further discussions.

Reply:
FYI https://devblogs.microsoft.com/oldnewthing/20140811-00/?p=283